File: android/app/src/main/java/com/fuggel/Uway/service/InstructionsManager.kt
package com.fuggel.Uway.service

import android.content.Context
import android.location.Location
import com.fuggel.Uway.model.Instruction
import com.fuggel.Uway.model.ManeuverType
import com.fuggel.Uway.model.VoiceInstruction
import com.fuggel.Uway.utils.GeoHelper
import com.fuggel.Uway.utils.NotificationHelper
import kotlin.math.abs

class InstructionsManager(
    private val steps: List<Instruction>,
    private val ttsManager: TTSManager,
    private val allowTextToSpeech: Boolean,
    private val getLastInstruction: () -> String?,
    private val setLastInstruction: (String) -> Unit,
    private val onArrived: () -> Unit,
    private val onDeviated: () -> Unit,
) {
    private val routeCoordinates: List<List<Double>> = steps.map { it.maneuverLocation }
    private val distanceThreshold = 50.0
    private val deviationThresholdMeters = 30.0
    private val uTurnAngleMin = 150.0
    private val uTurnAngleMax = 210.0

    private var isArrived = false

    fun updateLocation(location: Location, context: Context) {
        if (checkIfArrived(location)) {
            if (!isArrived) {
                isArrived = true
                onArrived()
            }
            return
        }

        if (checkDeviation(location)) {
            onDeviated()
        }

        val currentStep = getClosestStep(location) ?: return
        val distanceToManeuver = GeoHelper.haversine(
            location.latitude, location.longitude,
            currentStep.maneuverLocation[1], currentStep.maneuverLocation[0]
        )

        val activeVoice =
            getActiveVoiceInstruction(currentStep.voiceInstructions, distanceToManeuver)

        if (activeVoice != null && activeVoice.announcement != getLastInstruction()) {
            NotificationHelper.showNotification(context, "Navigation", activeVoice.announcement)

            setLastInstruction(activeVoice.announcement)

            if (allowTextToSpeech) {
                ttsManager.speak(activeVoice.announcement)
            }
        }
    }

    private fun checkDeviation(location: Location): Boolean {
        if (routeCoordinates.size < 2) return false

        val userPoint = doubleArrayOf(location.longitude, location.latitude)
        val (nearestPoint, indexOnLine, distance) = GeoHelper.nearestPointOnLine(
            routeCoordinates,
            userPoint
        )

        if (distance > deviationThresholdMeters) return true

        val nextIndex = (indexOnLine + 1).coerceAtMost(routeCoordinates.size - 1)
        val nextPoint = routeCoordinates[nextIndex]
        val routeBearing =
            GeoHelper.bearing(nearestPoint.toDoubleArray(), nextPoint.toDoubleArray())

        val userBearing = location.bearing.toDouble()
        val angleDiff = abs(routeBearing - userBearing)

        return angleDiff in uTurnAngleMin..uTurnAngleMax
    }

    private fun checkIfArrived(location: Location): Boolean {
        val lastStep = steps.lastOrNull() ?: return false
        if (lastStep.maneuverType != ManeuverType.ARRIVE) return false

        val (lng, lat) = lastStep.maneuverLocation
        val dist = GeoHelper.haversine(location.latitude, location.longitude, lat, lng)
        return dist <= 5.0
    }

    private fun getClosestStep(location: Location): Instruction? {
        val userLat = location.latitude
        val userLng = location.longitude

        return steps.minByOrNull {
            val (lng, lat) = it.maneuverLocation
            GeoHelper.haversine(userLat, userLng, lat, lng)
        }
    }

    private fun getActiveVoiceInstruction(
        voiceInstructions: List<VoiceInstruction>,
        distanceToManeuver: Double
    ): VoiceInstruction? {
        if (voiceInstructions.isEmpty()) return null

        val firstInstruction = voiceInstructions.first()

        return voiceInstructions.firstOrNull { instruction ->
            val diff = abs(distanceToManeuver - instruction.distanceAlongGeometry)
            val isAhead = instruction.distanceAlongGeometry >= distanceToManeuver
            val isFirst = instruction.announcement == firstInstruction.announcement
            isFirst || (isAhead && diff <= distanceThreshold)
        }
    }
}

File: ios/Services/InstructionsManager.swift
//
//  InstructionsManager.swift
//  Uway
//
//  Created by Furkan Ceylan on 19.06.25.
//

import Foundation
import CoreLocation

class InstructionsManager {
    private let steps: [Instruction]
    private let speak: (String) -> Void
    private let getLastInstruction: () -> String?
    private let setLastInstruction: (String) -> Void
    private let onArrived: () -> Void
    private let onDeviated: () -> Void
    
    private var isArrived = false
    private let routeCoordinates: [[Double]]
    
    init(
        steps: [Instruction],
        speak: @escaping (String) -> Void,
        getLastInstruction: @escaping () -> String?,
        setLastInstruction: @escaping (String) -> Void,
        onArrived: @escaping () -> Void,
        onDeviated: @escaping () -> Void
    ) {
        self.steps = steps
        self.speak = speak
        self.getLastInstruction = getLastInstruction
        self.setLastInstruction = setLastInstruction
        self.onArrived = onArrived
        self.onDeviated = onDeviated
        self.routeCoordinates = steps.map { $0.maneuverLocation }
    }
    
    func updateLocation(_ location: CLLocation) {
        guard !checkIfArrived(location: location) else {
            if !isArrived {
                isArrived = true
                print("navigation: destination reached")
                onArrived()
            }
            return
        }

        if checkDeviation(location: location) {
            print("navigation: route deviation detected")
            onDeviated()
        }

        guard let currentStep = getClosestStep(to: location) else {
            print("navigation: no matching step found")
            return
        }

        let distanceToManeuver = GeoHelper.haversine(
            lat1: location.coordinate.latitude,
            lon1: location.coordinate.longitude,
            lat2: currentStep.maneuverLocation[1],
            lon2: currentStep.maneuverLocation[0]
        )

        if let activeVoice = getActiveVoiceInstruction(
            voiceInstructions: currentStep.voiceInstructions,
            distanceToManeuver: distanceToManeuver
        ) {
            if activeVoice.announcement != getLastInstruction() {
                setLastInstruction(activeVoice.announcement)
                speak(activeVoice.announcement)
                NotificationHelper.showNotification(
                    title: "Navigation",
                    body: activeVoice.announcement
                )
            }
        }
    }

    private func checkIfArrived(location: CLLocation) -> Bool {
        guard let lastStep = steps.last,
              lastStep.maneuverType == .arrive else { return false }

        let (lng, lat) = (lastStep.maneuverLocation[0], lastStep.maneuverLocation[1])
        let dist = GeoHelper.haversine(
            lat1: location.coordinate.latitude,
            lon1: location.coordinate.longitude,
            lat2: lat,
            lon2: lng
        )
      
        return dist <= 5.0
    }

    private func checkDeviation(location: CLLocation) -> Bool {
        guard routeCoordinates.count >= 2 else { return false }

        let userPoint = [location.coordinate.longitude, location.coordinate.latitude]
        let result = GeoHelper.nearestPointOnLine(line: routeCoordinates, point: userPoint)

        if result.distanceMeters > Double(AppConfig.routeDeviationThresholdInMeters) {
            return true
        }

        guard location.course >= 0 else {
            return false
        }
      
        let nextIndex = min(result.indexOnLine + 1, routeCoordinates.count - 1)
        let nextPoint = routeCoordinates[nextIndex]
        let routeBearing = GeoHelper.bearing(
            from: result.nearestPoint,
            to: nextPoint
        )
        let userBearing = location.course
        let angleDiff = abs(routeBearing - userBearing)
      
        return angleDiff >= Double(AppConfig.uTurnAngleMin) && angleDiff <= Double(AppConfig.uTurnAngleMax)
    }

    private func getClosestStep(to location: CLLocation) -> Instruction? {
        steps.min(by: { lhs, rhs in
            let lhsDist = GeoHelper.haversine(
                lat1: location.coordinate.latitude,
                lon1: location.coordinate.longitude,
                lat2: lhs.maneuverLocation[1],
                lon2: lhs.maneuverLocation[0]
            )
            let rhsDist = GeoHelper.haversine(
                lat1: location.coordinate.latitude,
                lon1: location.coordinate.longitude,
                lat2: rhs.maneuverLocation[1],
                lon2: rhs.maneuverLocation[0]
            )
            return lhsDist < rhsDist
        })
    }

    private func getActiveVoiceInstruction(
        voiceInstructions: [VoiceInstruction],
        distanceToManeuver: Double
    ) -> VoiceInstruction? {
        guard !voiceInstructions.isEmpty else { return nil }
        let first = voiceInstructions.first!
        return voiceInstructions.first(where: {
            let diff = abs($0.distanceAlongGeometry - distanceToManeuver)
            let isAhead = $0.distanceAlongGeometry >= distanceToManeuver
            let result = $0.announcement == first.announcement || (isAhead && diff <= Double(AppConfig.distanceThresholdInMeters))
          
            return result
        })
    }
}

